packages/matter.js/src/protocol/ExchangeManager.ts

```diff
@@ -146,35 +146,95 @@ export class ExchangeManager<ContextT> {
+        let exchange = this.exchanges.get(exchangeIndex);
+
+        if (
+            exchange !== undefined &&
+            (exchange.session.getId() !== session.getId() ||
+                exchange.isInitiator === message.payloadHeader.isInitiatorMessage) // Should always be ok, but just in case
+        ) {
+            exchange = undefined;
+        }
+
         if (exchange !== undefined) {
-            await exchange.onMessageReceived(message);
+            await exchange.onMessageReceived(message, isDuplicate);
         } else {
             if (session.closingAfterExchangeFinished) {
                 throw new MatterFlowError(
                     `Session with ID ${packet.header.sessionId} marked for closure, decline new exchange creation.`,
                 );
             }
-            const exchange = MessageExchange.fromInitialMessage(
-                await this.channelManager.getOrCreateChannel(channel, session),
-                this.messageCounter,
-                message,
-                async () => await this.deleteExchange(exchangeIndex),
-            );
-            this.exchanges.set(exchangeIndex, exchange);
-            await exchange.onMessageReceived(message);
+
             const protocolHandler = this.protocols.get(message.payloadHeader.protocolId);
-            if (protocolHandler === undefined)
-                throw new MatterFlowError(`Unsupported protocol ${message.payloadHeader.protocolId}`);
-            await protocolHandler.onNewExchange(exchange, message);
+
+            if (protocolHandler !== undefined && message.payloadHeader.isInitiatorMessage && !isDuplicate) {
+                const exchange = MessageExchange.fromInitialMessage(
+                    await this.channelManager.getOrCreateChannel(channel, session),
+                    message,
+                    async () => await this.deleteExchange(exchangeIndex),
+                );
+                this.exchanges.set(exchangeIndex, exchange);
+                await exchange.onMessageReceived(message);
+                await protocolHandler.onNewExchange(exchange, message);
+            } else if (message.payloadHeader.requiresAck) {
+                const exchange = MessageExchange.fromInitialMessage(
+                    await this.channelManager.getOrCreateChannel(channel, session),
+                    message,
+                    async () => await this.deleteExchange(exchangeIndex),
+                );
+                this.exchanges.set(exchangeIndex, exchange);
+                await exchange.send(MessageType.StandaloneAck, new ByteArray(0));
+                await exchange.close();
+                logger.debug(
+                    `Ignoring unsolicited message ${messageId} for protocol ${message.payloadHeader.protocolId}.`,
+                );
+            } else {
+                if (protocolHandler === undefined)
+                    throw new MatterFlowError(`Unsupported protocol ${message.payloadHeader.protocolId}`);
```

Comment:
This is coming from the wire?  Maybe log a different message from below but as warning?  Don't like malformed wire messages to create what's effectively an internal error

Reply:
Yes and nmo ... ideally we should jhave all relevant ProtocolHandlers registered ... so when a mesage comes in with unexpected one could also be on us ...

Reply:
but I adjust to warning logging